src/modules/src/controller_indi.c

```diff
@@ -0,0 +1,311 @@
+ * It estimates the actuator effectiveness online, by comparing the expected
+ * angular acceleration based on the inputs with the measured angular
+ * acceleration
+ */
+static inline void lms_estimation(float stateAttitudeRateRoll, float stateAttitudeRatePitch, float stateAttitudeRateYaw)
+{
+  static struct IndiEstimation *est = &indi.est;
+  // Only pass really low frequencies so you don't adapt to noise
+  struct FloatRates body_rates = {
+			  .p = stateAttitudeRateRoll,
+			  .q = stateAttitudeRatePitch,
+			  .r = stateAttitudeRateYaw,
+  };
+  filter_pqr(est->u, &indi.u_act_dyn);
+  filter_pqr(est->rate, &body_rates);
+
+  // Calculate the first and second derivatives of the rates and actuators
+  float rate_d_prev[3];
+  float u_d_prev[3];
+  float_vect_copy(rate_d_prev, est->rate_d, 3);
+  float_vect_copy(u_d_prev, est->u_d, 3);
+  finite_difference_from_filter(est->rate_d, est->rate);
+  finite_difference_from_filter(est->u_d, est->u);
+  finite_difference(est->rate_dd, est->rate_d, rate_d_prev);
+  finite_difference(est->u_dd, est->u_d, u_d_prev);
+
+  // The inputs are scaled in order to avoid overflows
+  float du[3];
+  float_vect_copy(du, est->u_d, 3);
+  float_vect_scale(du, INDI_EST_SCALE, 3);
+  est->g1.p = est->g1.p - (est->g1.p * du[0] - est->rate_dd[0]) * du[0] * est->mu;
+  est->g1.q = est->g1.q - (est->g1.q * du[1] - est->rate_dd[1]) * du[1] * est->mu;
+  float ddu = est->u_dd[2] * INDI_EST_SCALE / PERIODIC_FREQUENCY;
+  float error = (est->g1.r * du[2] + est->g2 * ddu - est->rate_dd[2]);
+  est->g1.r = est->g1.r - error * du[2] * est->mu / 3.0f;
+  est->g2 = est->g2 - error * 1000.0f * ddu * est->mu / 3.0f;
+
+  //the g values should be larger than zero, otherwise there is positive feedback, the command will go to max and there is nothing to learn anymore...
+  if (est->g1.p < 0.01f) { est->g1.p = 0.01f; }
+  if (est->g1.q < 0.01f) { est->g1.q = 0.01f; }
+  if (est->g1.r < 0.01f) { est->g1.r = 0.01f; }
+  if (est->g2   < 0.01f) { est->g2 = 0.01f; }
+
+  if (indi.adaptive) {
+    //Commit the estimated G values and apply the scaling
+    indi.g1.p = est->g1.p * INDI_EST_SCALE;
+    indi.g1.q = est->g1.q * INDI_EST_SCALE;
+    indi.g1.r = est->g1.r * INDI_EST_SCALE;
+    indi.g2   = est->g2 * INDI_EST_SCALE;
+  }
+}
+
+void controllerINDIInit(void)
+{
+	float_rates_zero(&indi.angular_accel_ref);
+	float_rates_zero(&indi.u_act_dyn);
+	float_rates_zero(&indi.u_in);
+
+	// Re-initialize filters
+	indi_init_filters();
+}
+
+bool controllerINDITest(void)
+{
+	return true;
+}
+
+void controllerINDI(control_t *control, setpoint_t *setpoint,
+                                         const sensorData_t *sensors,
+                                         const state_t *state,
+                                         const uint32_t tick)
+{
+
+	/*
+	 * 1 - Update the gyro filter with the new measurements.
+	 */
+
+	float stateAttitudeRateRoll = radians(sensors->gyro.x);
+	float stateAttitudeRatePitch = -radians(sensors->gyro.y);
```

Comment:
Are you sure about this minus?

Reply:
Yes, I am sure. There is this caveat in the Crazyflie that the Y Axis is inverted. It is documented and used like that in all of their controllers. It is something we have to be aware of all the time not to forget.

Reply:
ok!